Fix havocing of arrays when enforcing invariants

SaswatPadhi opened the pull request:
Fixes #6020.

As reported in #6020, only the first element of an array was being havoced earlier. In this change, we fix this behavior using `havoc_object`.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- NA ~~The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/~~
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- NA ~~My commit message includes data points confirming performance improvements (if claimed).~~
- [x] My PR is restricted to a single feature or bugfix.
- NA ~~White-space or formatting changes outside the feature-related changed lines are in commits of their own.~~

SaswatPadhi committed:
Fix havocing of arrays when enforcing invariants

As reported in #6020, only the first element of an array was being
havoced earlier.

In this change, we fix this behavior using `havoc_object`.

tautschnig commented:
@SaswatPadhi Please update `scripts/expected_doxygen_warnings.txt` to silence the `doxygen` warning.

tautschnig commented on regression/contracts/invar_havoc_dynamic_multi-dim_array_partial_const_idx/test.desc:
```diff
@@ -0,0 +1,21 @@
+KNOWNBUG
```
I'd prefer to have all the `KNOWNBUG` tests in a separate commit, especially as they (also) relate to a different issue.

SaswatPadhi replied:
Good point! Done.

tautschnig commented on src/goto-instrument/loop_utils.cpp:
```diff
@@ -35,23 +58,51 @@ goto_programt::targett get_loop_exit(const loopt &loop)
+  side_effect_expr_nondett rhs(lhs.type(), loop_head->source_location);
+
+  goto_programt::targett t = dest.add(goto_programt::make_assignment(
+    code_assignt(std::move(lhs), std::move(rhs)), loop_head->source_location));
```
`lhs` is `const &`, the `std::move` will not add value. Also, you can nowadays use `make_assignment(lhs, std::move(rhs), loop_head->source_location)` (avoiding the `code_assignt`).

SaswatPadhi replied:
Thanks! Fixed.

tautschnig commented on src/goto-instrument/loop_utils.cpp:
```diff
@@ -35,23 +58,51 @@ goto_programt::targett get_loop_exit(const loopt &loop)
+
+  goto_programt::targett t =
+    dest.add(goto_programt::make_other(havoc, loop_head->source_location));
+  t->code_nonconst().add_source_location() = loop_head->source_location;
```
This is redundant, line 79 already sets that source location.

tautschnig commented on src/goto-instrument/loop_utils.cpp:
```diff
@@ -35,23 +58,51 @@ goto_programt::targett get_loop_exit(const loopt &loop)
-    side_effect_expr_nondett rhs(lhs.type(), loop_head->source_location);
+    if(lhs.id() == ID_index || lhs.id() == ID_dereference)
+    {
+      address_of_exprt address_of_lhs(std::move(lhs));
```
`lhs` is `const &`, no need for `std::move`.

tautschnig commented on src/goto-instrument/loop_utils.cpp:
```diff
@@ -63,17 +114,23 @@ void get_modifies_lhs(
   if(lhs.id()==ID_symbol)
     modifies.insert(lhs);
-  else if(lhs.id()==ID_dereference)
+  else if(lhs.id() == ID_dereference)
```
Nit pick: I'd leave this line unchanged.

tautschnig commented on src/goto-instrument/loop_utils.cpp:
```diff
@@ -63,17 +114,23 @@ void get_modifies_lhs(
   {
   }
-  else if(lhs.id()==ID_index)
+  else if(lhs.id() == ID_index)
```
Nit pick: I'd leave this line unchanged.

SaswatPadhi committed:
Update `expected_doxygen_warnings.txt`

SaswatPadhi committed:
More regression tests for havocing dynamic arrays

These regression tests currently fail due to imprecision in alias
analysis (see: #6021).

In future, we could either improve the alias analysis, or add manual
assigns clause annotations on these loops and make sure that the arrays
are correctly havoced.

chrisr-diffblue commented on src/goto-instrument/loop_utils.cpp:
```diff
@@ -11,12 +11,35 @@ Author: Daniel Kroening, kroening@kroening.com
+  explicit loop_utils_is_constantt(const modifiest &mod) : modifies(mod)
+  {
+  }
+  bool is_constant(const exprt &expr) const override
```
Minor query - I wonder if this would be better with a more explicit name, such as `is_constant_or_unmodified` ?

SaswatPadhi replied:
I agree, that would be a better name. But the `operator()` defined on the base class (`is_constantt`) calls `is_constant` and so I chose to override the `is_constant` function.

chrisr-diffblue replied:
Yeah, I assumed it was a "more hassle than its worth" choice :-)

chrisr-diffblue reviewed (approved):
Looks good, and happy for it to go in - just a minor query about naming, which you can feel free to ignore if it gets too complicated.
